### PR TITLE
[dashboard] Always start with options

### LINKED ISSUE #16362: Start workspaces using the currently selected org
We should respect and show the currently selected org when starting a workspace. The org should also be an explicit argument to the createWorkspace call.

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
Introduces a new CreateWorkspacePage on which the options are shown and can be changed.

The page is only activated for users who are in team-only mode already (change your email to `toa@gitpod.io` in user settings and reload the dashboard to activate it). There is also a feature flag that can enable it for other users (currently false everywhere). 

This also moves /new to /projects/new so /new can be used for creating new workspaces.

**Without the feature flag there will be no difference in starting workspaces, with one exception: `/new` works for everybody**

## Related Issue(s)

Fixes #16362

## How to test

1) Try starting workspaces in different ways. Should all work as usual.
2) Change the user email to `toa@gitpod.io` and reload the browser to active org only attribution.
3) All workspace creations now go to a new create workspace page. 

## Release Notes

```release-note
New workspace creation page
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-slow-database
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@svenefftinge: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:153]: **question**: Also, is it better to move the loading state out the button instead of showing the button (already) disabled with a spinner and changing the label to creating workspace?

| BEFORE | AFTER :a: | AFTER :b: |
|-|-|-|
| new-before | new-after-1 | new-after-2 |

File: components/dashboard/src/projects/projects.routes.ts
Comment (gtsiolis) [inline components/dashboard/src/projects/projects.routes.ts:14]: **issue**: The new project button on the projects empty state and project list actions still links back top `/new`.

File: components/dashboard/src/app/AppRoutes.tsx
Comment (gtsiolis) [inline components/dashboard/src/app/AppRoutes.tsx:163]: **suggestion**: Shall we drop this route now as it seems duplicate and broken? See [relevant discussion](https://gitpod.slack.com/archives/C01KPEPQYE6/p1673523784907079?thread_ts=1672836161.790069&cid=C01KPEPQYE6) (internal).

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (gtsiolis) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:126]: **nitpick**: Minor heading style fix.

```suggestion
                <p className="text-gray-500 text-center text-base">Start a new workspace with the following options.</p>
```

**Review by gtsiolis — APPROVED** [review]
UX LGTM! 🌮 🌮 

Left some comments below.

❓ What is `toa@gitpod.io`? I changed it in user settings, but didn't notice any difference—the /new page and flow remained the same. Should I see something different?

Approving to unblock merging but holding for someone from `@gitpod-io/engineering-webapp`  to take a closer look at the code changes. 🏀 

/hold

**Comment by svenefftinge** [comment]
> What is toa@gitpod.io?

It enables the **t**eam_**o**nly_**a**ttribution feature flag.
You need to reload the dashboard after that change.

**Comment by atduarte** [comment]
Looking good! 🥳 

Had a few issues though. I visited "/new" and at first it showed me a "Oh no something went wrong". Retrying worked. Same happened when I used the "gitpodl-link#repository-url" trick at first. And the second time the options did not appear. Am I doing something wrong?

Also noticed that the loading animation on the "Creating workspace" button had some issues.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:31]: Curious on why we want to connect the user team-only migration to this new flow?

File: components/dashboard/src/projects/projects.routes.ts
Comment (selfcontained) [inline components/dashboard/src/projects/projects.routes.ts:14]: This route makes so much more sense for creating a new project than just `/new` did - thanks 😄

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:39]: This could be more explicit if we didn't allow it to be `undefined` and coerced into a `boolean`.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:60]: nit: can drop the `<boolean>` type here - it will be inferred from the `false` default passed in.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:85]: With this being new code, it would be really nice to use a react query mutation for this. If you don't have time, maybe we could drop a comment, and I can follow up with moving it there. It'll help with adding consistency for managing things like the creating/saving/error states of async api calls like this.

There are a couple examples in the `src/data` directory if you end up wanting to give it a shot though. There's already a few workspace related queries/mutations present, so could add a new `data/workspaces/create-workspace-mutation.ts`. A pretty simple example for reference exists in `data/workspaces/stop-workspace-mutation.ts`. An extra mile could be invalidating the list workspaces query in an `onSuccess` handler for the new mutation. It'd look something like:

```ts
onSuccess: () => {
    const queryKey = getListWorkspacesQueryKey();
    queryClient.invalidateQueries({ queryKey });
}
```

Then it'd remove the need to manually manage a few pieces of state in this component, namely `creating`, `createError`, `createResult`.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:101]: I think this is an unnecessary wrapper around `createWorkspace`, as it will change whenever `createWorkspace` changes, and just calls it. We could just pass `createWorkspace` below where we're passing in `onClickCreate`

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:113]: Could we shift this out of the render loop and into the `createWorkspace` function right after we get the `result` from the api call? It would be much more explicit that we're navigating after creation if there's a workspaceUrl on the result. It would also remove the need to have component state for that result.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:117]: likewise I think this would be better handled in the `createWorkspace` function after we get the result

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:119]: Instead of relying on the result for this check here, I think it would be clearer if we had something like a `showExistingWorkspaceModal` component state, and we just set that in the `createWorkspace()` function if this condition happens, and render the modal based only on that boolean flag.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:155]: related to above, could this just be `onClick={createWorkspace}` instead?

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:152]: I really like the UX of having a button that supports a `loading` state w/ a spinner. I think it'd be great for us to extract this into a proper `<Button/>` component that supports things like a `loading` prop that shows this spinner. Could we add a todo here and perhaps it's something I could extract when I have time as a followup?

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:206]: A suggestion for a slightly different way to define a component that handles the return type automatically. You're handling a return type of `JSX.Element` here, so it won't make a big difference, but it's easy to forget the return type, and then some lint errors aren't caught.

```ts
type StatusMessageProps = {...}
const StatusMessage: FunctionComponent<StatusMessageProps> = ({ ... }) => {
```

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:292]: Ditto on the comment above about using `FunctionComponent<Props>` for defining the component. In this case we're missing the return type, so there are some linter errors that wouldn't get caught.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:315]: We should add a `key` to the `<a>` here - any array of items rendered (usually in a `.map()` call) need a unique `key`. I'm not sure if there's a reliable unique property of `w` in this loop, but if so, that's a good value to use for `key`.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/CreateWorkspacePage.tsx:317]: What's the value of using `gitpodHostUrl()` here vs. just using a value like `/start/#${w.latestInstance?.workspaceId}`? I see this in various places, and it doesn't seem like we need a full absolute url and could simplify this.

File: components/dashboard/src/workspaces/StartWorkspaceModal.tsx
Comment (selfcontained) [inline components/dashboard/src/workspaces/StartWorkspaceModal.tsx:64]: Is the intent of putting this inside `StartWorkspaceModal` to make it easier to make all the existing places that start a workspace to run through here? Do you think we'd eventually be able to remove this `StartWorkspaceModal` and replace the locations opening this modal with just a link to the new page?

**Review by selfcontained — COMMENTED** [review]
Haven't been able to test it yet - preview env isn't loading for me atm - but wanted to share some comments on just the changes w/o waiting for testing it out.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (svenefftinge) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:31]: User's who are in team only mode, will no longer have the cost center setting in the user settings. So it's even more important for them to be clear in which org the workspace is started.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (svenefftinge) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:85]: Not sure I understand for what you want to use a mutation. Can you clarify?

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (svenefftinge) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:101]: The signature doesn't fit the expected click event, because both accepot a different first argument.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (svenefftinge) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:113]: Not sure I understood your suggestion, the navigation only happens on close of the Modal which needs to be rendered.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (svenefftinge) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:155]: unfortunately not, see my comment above.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (gtsiolis) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:155]: +1 — We could replace about 60 button instances across the product. Opened https://github.com/gitpod-io/gitpod/issues/16441 to keep track of this.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:87]: Mutation is a term react-query uses for how it provides an abstraction for any api calls that mutate data, as opposed to a query, that just fetches data.

For example, [here's a](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/data/workspaces/stop-workspace-mutation.ts) "mutation" for the stop workspace api call. It's basically a hook wrapper around react-query's `useMutation()` hook.

[Here's what](https://github.com/gitpod-io/gitpod/blob/facc352acf6821c185553e2d8bb7ea4b859f1c17/components/dashboard/src/workspaces/WorkspaceOverflowMenu.tsx#L34-L50) it looks like when that mutation is used in a component.

In the case here, we actually want to await the mutation (rather than fire and forget), so it'd look slightly different, we'd use the `mutateAsync` fn like [here](https://github.com/gitpod-io/gitpod/blob/64d46a9aacb2e563ac436aa527f1bdb9df631a72/components/dashboard/src/teams/git-integrations/GitIntegrationListItem.tsx#L38-L45).

The "mutation" return value from calling the hook has some properties, like `isLoading`, `isError`, `data` (the returned result), and `error` if an error was thrown. These can be used in place of manually managing that state in the component [like here](https://github.com/gitpod-io/gitpod/blob/64d46a9aacb2e563ac436aa527f1bdb9df631a72/components/dashboard/src/teams/git-integrations/GitIntegrationListItem.tsx#L77-L78).

There's a lot more they can do that we can take advantage of over time. If you're curious, here's the full docs. Also, This would be something I would be happy to refactor as a followup, but wanted to mention it, as I think it'd be great if our new stuff could take advantage of react-query.

https://tanstack.com/query/v4/docs/react/guides/mutations

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:116]: `createWorkspace` could look like something like this, and use a more explicit imperative approach that doesn't have the indirection of setting a result as state, and expecting a re-render of the component to then trigger a navigation. I was wrong about the `ExistingWorkspaceModal` though, didn't realize it needs the result to render, but we could still make it a bit more clear I think:

```ts
// In createWorkspace()
const result = await getGitpodService().server.createWorkspace({
    contextUrl: repo,
    organizationId: team?.id,
    ...opts,
});

// handle navigation here
if(result?.workspaceURL) {
    window.location.href = result.workspaceURL;
    return;
}

// and here
if (result?.createdWorkspaceId) {
    history.replace(`/start/${createResult.createdWorkspaceId}`);
    return;
}

// This one is debatable, I didn't realize the modal render needs the result, but this could be more clear by something like...
if (result?.existingWorkspaces) {
    setExistingWorkspaceResult(result)
    return;
}

//... then in the rendering
if (existingWorkspaceResult) {
    return (
        <ExistingWorkspaceModal 
            result={existingWorkspaceResult} 
            createWorkspace={createWorkspace} onClose={() => { }} 
        />
    )
}
```

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (selfcontained) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:122]: I was wrong here, it needs the result to render, more details in my response above.

File: components/dashboard/src/workspaces/CreateWorkspacePage.tsx
Comment (svenefftinge) [reply components/dashboard/src/workspaces/CreateWorkspacePage.tsx:87]: Would be great if you could do that refactoring in a follow up. Thanks!

File: components/dashboard/src/workspaces/StartWorkspaceModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/workspaces/StartWorkspaceModal.tsx:64]: Yes, it's here as kind of a catch all until we have fully rolled out the new CreateWorkspaceFlow.

**Pull request merged**
